improver/calibration/rainforest_calibration.py

```diff
@@ -419,13 +422,33 @@ def _calculate_error_probabilities(
         """
         error_probability_cube = self._prepare_error_probability_cube(forecast_cube)
 
-        input_dataset = self._prepare_features_dataframe(feature_cubes)
+        input_dataset, feature_variables = self._prepare_features_array(feature_cubes)
+
+        precipitation_ind = feature_variables.index(
+            "lwe_thickness_of_precipitation_amount"
+        )
+        precip_forecast = input_dataset[:, precipitation_ind]
 
         for threshold_index, model in enumerate(self.tree_models):
-            prediction = model.predict(input_dataset)
-            error_probability_cube.data[threshold_index, ...] = np.reshape(
-                prediction, forecast_cube.data.shape
-            )
+            threshold = self.error_thresholds[threshold_index]
+            if threshold >= 0:
+                prediction = model.predict(input_dataset)
+                error_probability_cube.data[threshold_index, ...] = np.reshape(
+                    prediction, forecast_cube.data.shape
+                )
+            else:
+                forecast_bool = precip_forecast + threshold > 0
+                if np.any(forecast_bool):
+                    prediction = np.ones(input_dataset.shape[0], dtype=np.float32)
+                    input_subset = input_dataset[forecast_bool]
+                    prediction[forecast_bool] = model.predict(input_subset)
+                    error_probability_cube.data[threshold_index, ...] = np.reshape(
+                        prediction, forecast_cube.data.shape
+                    )
+                else:
+                    error_probability_cube.data[threshold_index, ...] = np.ones(
+                        forecast_cube.data.shape
+                    )
```

Comment:
I've refactored the code along the lines you suggested. I extracted the whole loop over thresholds, rather than just the inner part, because in the treelite case, we want to reuse the same prediction data DMatrix for all thresholds >= 0 (creating it repeatedly from the input array would be inefficient), so we'd need to pass it in along with the array, which seemed too cumbersome.